tracer/src/Datadog.Trace/Configuration/TracerSettings.cs

```diff
@@ -383,19 +384,33 @@ public ImmutableTracerSettings Build()
             return new ImmutableTracerSettings(this);
         }
 
-        private static IDictionary<string, string> InitializeHeaderTags(IDictionary<string, string> configurationDictionary)
+        private static IDictionary<string, string> InitializeHeaderTags(IDictionary<string, string> configurationDictionary, bool headerTagsNormalizationFixEnabled)
         {
             var headerTags = new Dictionary<string, string>();
 
-            foreach (KeyValuePair<string, string> kvp in configurationDictionary)
+            foreach (var kvp in configurationDictionary)
             {
-                if (!string.IsNullOrWhiteSpace(kvp.Key) && string.IsNullOrWhiteSpace(kvp.Value))
+                var headerName = kvp.Key;
+                var providedTagName = kvp.Value;
+                if (string.IsNullOrWhiteSpace(headerName))
+                {
+                    continue;
+                }
+
+                // The user has not provided a tag name. The normalization will happen later, when adding the prefix.
+                if (string.IsNullOrEmpty(providedTagName))
+                {
+                    headerTags.Add(headerName.Trim(), string.Empty);
+                }
+                else if (headerTagsNormalizationFixEnabled && providedTagName.TryConvertToNormalizedTagName(normalizePeriods: false, out var normalizedTagName))
                 {
-                    headerTags.Add(kvp.Key.Trim(), string.Empty);
+                    // If the user has provided a tag name, then we don't normalize periods in the provided tag name
+                    headerTags.Add(headerName.Trim(), normalizedTagName);
                 }
-                else if (!string.IsNullOrWhiteSpace(kvp.Key) && kvp.Value.TryConvertToNormalizedHeaderTagName(out string result))
+                else if (!headerTagsNormalizationFixEnabled && providedTagName.TryConvertToNormalizedTagName(normalizePeriods: true, out var normalizedTagNameNoPeriods))
                 {
-                    headerTags.Add(kvp.Key.Trim(), result);
+                    // Back to the previous behaviour if the flag is set
+                    headerTags.Add(headerName.Trim(), normalizedTagNameNoPeriods);
                 }
```

Comment:
Maybe not, but is it worth logging if they _have_ provided a value, but the normalization fails due to the tag being too long/not starting with a letter? 

EDIT: Actually, I'm not sure it's safe to log here yet, so maybe not

Reply:
I asked myself the question but at the time I thought that the documentation is clear and it could be something for the CLI tool to check rather than for us to add logs here.

I would never have guessed that logging here would not be safe. TBH, I don't see why. We even instantiate a logger in `TracerManagerFactory` (ok used a bit later but still instantiated earlier), so I would have said it would be ok. If you have more insights on this, I'd love to understand.

Reply:
Merging now, as I need to port it to 1.x but can revisit if need be

Reply:
> I would never have guessed that logging here would not be safe. If you have more insights on this, I'd love to understand.

I didn't look into it very hard 😄 I just thought "where does the logger get it's configuration from". Turns out it's entirely from env vars, so it's fine 😉 Definitely not a priority to add the log IMO, was just a thought